guacamole/src/main/java/org/apache/guacamole/tunnel/InterceptedStream.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.guacamole.protocol.GuacamoleStatus;
+
+/**
+ * A simple pairing of the index of an intercepted Guacamole stream with the
+ * stream-type object which will produce or consume the data sent over the
+ * intercepted Guacamole stream.
+ *
+ * @author Michael Jumper
+ * @param <T>
+ *     The type of object which will produce or consume the data sent over the
+ *     intercepted Guacamole stream. Usually, this will be either InputStream
+ *     or OutputStream.
+ */
+public class InterceptedStream<T extends Closeable> {
+
+    /**
+     * The index of the Guacamole stream being intercepted.
+     */
+    private final String index;
+
+    /**
+     * The stream which will produce or consume the data sent over the
+     * intercepted Guacamole stream.
+     */
+    private final T stream;
+
+    /**
+     * The exception which prevented the stream from completing successfully,
+     * if any. If the stream completed successfully, or has not encountered any
+     * exception yet, this will be null.
+     */
+    private GuacamoleException streamError = null;
+
+    /**
+     * Creates a new InterceptedStream which associated the given Guacamole
+     * stream index with the given stream object.
+     *
+     * @param index
+     *     The index of the Guacamole stream being intercepted.
+     *
+     * @param stream
+     *     The stream which will produce or consume the data sent over the
+     *     intercepted Guacamole stream.
+     */
+    public InterceptedStream(String index, T stream) {
+        this.index = index;
+        this.stream = stream;
+    }
+
+    /**
+     * Returns the index of the Guacamole stream being intercepted.
+     *
+     * @return
+     *     The index of the Guacamole stream being intercepted.
+     */
+    public String getIndex() {
+        return index;
+    }
+
+    /**
+     * Returns the stream which will produce or consume the data sent over the
+     * intercepted Guacamole stream.
+     *
+     * @return
+     *     The stream which will produce or consume the data sent over the
+     *     intercepted Guacamole stream.
+     */
+    public T getStream() {
+        return stream;
+    }
+
+    /**
+     * Reports that this InterceptedStream did not complete successfully due to
+     * the given GuacamoleException, which could not be thrown at the time due
+     * to asynchronous handling of the stream contents.
+     *
+     * @param streamError
+     *     The exception which prevented the stream from completing
+     *     successfully.
+     */
+    public void setStreamError(GuacamoleException streamError) {
+        this.streamError = streamError;
+    }
+
+    /**
+     * Reports that this InterceptedStream did not complete successfully due to
+     * an error described by the given status code and human-readable message.
+     * The error reported by this call can later be retrieved as a
+     * GuacamoleStreamException by calling getStreamError().
```

Comment:
Since `GuacamoleStreamException` represents a received "ack" instruction, it can really only apply to streams where we are acting as the source of data and the server is sending acknowledgements ("ack" is only sent in response to "blob" or the instructions which start a stream). In practical terms of this change, if we changed `GuacamoleException` to `GuacamoleStreamException` for `setStreamError()` in general, that would mean that `setStreamError()` could only be used in context of `InputStreamInterceptingFilter`. If anything else ever wants to report an error, it will be impossible.

It is true that this is already the case - only `InputStreamInterceptingFilter` uses this function, but narrowing the type to that extent would guarantee that it can never be used anywhere else, and I felt that would be bad.